### PR TITLE
Protocol changes for R4.1

### PR DESCRIPTION
**Pull request by marmarek**
- Increase data chunk size to 64k.
- Increase allowed length of requested qrexec service name - MSG_TRIGGER_SERVICE message. 

Fixes https://github.com/QubesOS/qubes-issues/issues/4909

### CODE REVIEW

**Comment by marmarek** [comment]
This PR could use some extra pair of eyes on review, both on design level and implementation level.
cc @jpouellet @HW42

**Comment by jpouellet** [comment]
I've started to look, but I need to refresh my brain of the surrounding subsystems first (and the changes since I've last looked) to have a deep enough understanding and sufficient context in order to provide a meaningful review. It might be a little while.

File: agent/qrexec-agent-data.c [hunk 1]
```diff
@@ -152,17 +156,18 @@ void send_exit_code(libvchan_t *data_vchan, int status)
  *  1 - some data processed, call it again when buffer space and more data
  *      available
  */
-int handle_input(libvchan_t *vchan, int fd, int msg_type)
+int handle_input(libvchan_t *vchan, int fd, int msg_type, int data_protocol_version)
 {
-    char buf[MAX_DATA_CHUNK];
-    int len;
+    const size_t max_len = data_protocol_version < 3 ? MAX_DATA_CHUNK_V2 : MAX_DATA_CHUNK_V3;
+    char buf[max_len];
```
Comment (HW42) [inline agent/qrexec-agent-data.c]: I have seen that some code bases try to avoid allocating bigger chunks on the stack, and often avoid dynamic stack allocation altogether. Since the size has a fixed maximum it's probably ok as is.

File: agent/qrexec-agent-data.c [hunk 2]
```diff
@@ -152,17 +156,18 @@ void send_exit_code(libvchan_t *data_vchan, int status)
...
+    ssize_t len;
     struct msg_header hdr;
 
     hdr.type = msg_type;
     while (libvchan_buffer_space(vchan) > (int)sizeof(struct msg_header)) {
         len = libvchan_buffer_space(vchan)-sizeof(struct msg_header);
-        if (len > (int)sizeof(buf))
-            len = sizeof(buf);
+        if (len > (int)max_len)
```
Comment (HW42) [inline agent/qrexec-agent-data.c]: cast to int is wrong. Probably we want `(size_t)len > max_len`? (The mixing of `size_t` and `ssize_t` in functions like `read` is really annoying)

File: agent/qrexec-agent-data.c [hunk 3]
```diff
@@ -220,9 +226,9 @@ int handle_remote_data(libvchan_t *data_vchan, int stdin_fd, int *status,
     while (libvchan_data_ready(data_vchan) > 0) {
         if (libvchan_recv(data_vchan, &hdr, sizeof(hdr)) < 0)
             return -1;
-        if (hdr.len > MAX_DATA_CHUNK) {
-            fprintf(stderr, "Too big data chunk received: %d > %d\n",
-                    hdr.len, MAX_DATA_CHUNK);
+        if (hdr.len > max_len) {
+            fprintf(stderr, "Too big data chunk received: %d > %lu\n",
```
Comment (HW42) [inline agent/qrexec-agent-data.c]: `hdr.len` is `uint32_t` and `max_len` is `size_t` so `%" PRIu32 "` and `%zu` are the correct formats, I think.

File: agent/qrexec-agent.c [hunk 4]
```diff
@@ -868,28 +868,51 @@ int fill_fds_for_select(fd_set * rdset, fd_set * wrset)
...
-        if (libvchan_send(ctrl_vchan, &params, sizeof(params)) < 0)
-            handle_vchan_error("write params");
-    }
-    if (ret <= 0) {
-        close(client_fd);
+    if (!read_all(client_fd, &hdr, sizeof(hdr)))
+        goto error;
+    if (hdr.type != MSG_TRIGGER_SERVICE3 ||
+            hdr.len < sizeof(params) ||
+            hdr.len > MAX_SERVICE_NAME_LEN) {
```
Comment (HW42) [inline agent/qrexec-agent.c]: I think you want `hdr.len > sizeof(params) + MAX_SERVICE_LEN` here. This would also match the code in `sanitize_message_from_agent`.

File: agent/qrexec-agent.c [hunk 5]
```diff
@@ -868,28 +868,51 @@ int fill_fds_for_select(fd_set * rdset, fd_set * wrset)
...
+    if (!read_all(client_fd, &params, sizeof(params)))
+        goto error;
+    command_len = hdr.len - sizeof(params);
+    command = malloc(command_len);
+    if (!command)
+        goto error;
+    if (!read_all(client_fd, command, command_len))
+        goto error;
+    /* check '\0' at the end */
+    if (command[command_len-1])
```
Comment (HW42) [inline agent/qrexec-agent.c]: `if (command[command_len-1] != '\0')` and then drop redundant comment.

File: agent/qrexec-agent.c [hunk 6]
```diff
@@ -868,28 +868,51 @@ int fill_fds_for_select(fd_set * rdset, fd_set * wrset)
...
+        goto error;
+
+    snprintf(params.request_id.ident, sizeof(params.request_id), "SOCKET%d", client_fd);
+    if (libvchan_send(ctrl_vchan, &hdr, sizeof(hdr)) < 0)
+        handle_vchan_error("write hdr");
+    if (libvchan_send(ctrl_vchan, &params, sizeof(params)) < 0)
+        handle_vchan_error("write params");
+    if (libvchan_send(ctrl_vchan, command, command_len) < 0)
+        handle_vchan_error("write params");
```
Comment (HW42) [inline agent/qrexec-agent.c]: `"write command"`

File: agent/qrexec-client-vm.c [hunk 7]
```diff
@@ -132,10 +135,16 @@ int main(int argc, char **argv)
         start_local_process = 1;
     }
 
+    service_name = argv[optind + 1];
+
+    service_name_len_plus_one = strlen(service_name) + 1;
```
Comment (HW42) [inline agent/qrexec-client-vm.c]: All other code here seems to use length as a synonym for size (as supposed to length == strlen(x) == size - 1 as seen in some other code). So this should be just `service_name_len`.

File: daemon/qrexec-client.c [hunk 8]
```diff
@@ -40,6 +40,10 @@ int replace_chars_stderr = 0;
 
 #define VCHAN_BUFFER_SIZE 65536
 
+#if MAX_DATA_CHUNK_V3 < MAX_DATA_CHUNK_V2
+#error "MAX_DATA_CHUNK_V3 must not be smaller than MAX_DATA_CHUNK_V2"
+#endif
```
Comment (HW42) [inline daemon/qrexec-client.c]: Do wee need this check twice? Maybe move it into the header?

File: daemon/qrexec-client.c [hunk 9]
```diff
@@ -70,7 +76,9 @@ static int handle_agent_handshake(libvchan_t *vchan, int remote_send_first)
                 return -1;
             }
 
-            if (info.version != QREXEC_PROTOCOL_VERSION) {
+            data_protocol_version = info.version < QREXEC_PROTOCOL_VERSION ?
+                                    info.version : QREXEC_PROTOCOL_VERSION;
+            if (data_protocol_version < 2) {
```
Comment (HW42) [inline daemon/qrexec-client.c]: Formally hardcoded constants are against the style guide. I personally think this is fine but maybe consider to use something like `QREXEC_MIN_VERSION` or so.

There are a few other cases of hardcoded 2s and 3s. I won't tag them all.

File: daemon/qrexec-client.c [hunk 10]
```diff
@@ -280,16 +288,21 @@ static void send_exit_code(libvchan_t *vchan, int status)
     }
 }
 
-static void handle_input(libvchan_t *vchan)
+static void handle_input(libvchan_t *vchan, int data_protocol_version)
 {
-    char buf[MAX_DATA_CHUNK];
+    char buf[MAX_DATA_CHUNK_V3];
```
Comment (HW42) [inline daemon/qrexec-client.c]: Why do you allocate here always for V3?

File: daemon/qrexec-client.c [hunk 11]
```diff
@@ -280,16 +288,21 @@ static void send_exit_code(libvchan_t *vchan, int status)
...
     int ret;
     size_t max_len;
     struct msg_header hdr;
 
-    max_len = libvchan_buffer_space(vchan)-sizeof(hdr);
+    max_len = libvchan_buffer_space(vchan);
+    if (max_len < sizeof(hdr))
+        return;
```
Comment (HW42) [inline daemon/qrexec-client.c]: Don't we want some hard exit in this strange case?

File: daemon/qrexec-client.c [hunk 12]
```diff
@@ -357,11 +370,12 @@ void do_replace_chars(char *buf, int len) {
...
 }
 
-static int handle_vchan_data(libvchan_t *vchan, struct buffer *stdin_buf)
+static int handle_vchan_data(libvchan_t *vchan, struct buffer *stdin_buf,
+        int data_protocol_version)
 {
     int status;
     struct msg_header hdr;
-    char buf[MAX_DATA_CHUNK];
+    char buf[MAX_DATA_CHUNK_V3];
```
Comment (HW42) [inline daemon/qrexec-client.c]: Why always V3?

File: daemon/qrexec-daemon.c [hunk 13]
```diff
@@ -553,7 +556,7 @@ static void handle_client_hello(int fd)
 /* handle data received from one of qrexec_client processes */
 static void handle_message_from_client(int fd)
 {
-    char buf[MAX_DATA_CHUNK];
+    char buf[1];
```
Comment (HW42) [inline daemon/qrexec-daemon.c]: This change seems unrelated. Probably better moved to a separate commit.

File: daemon/qrexec-daemon.c [hunk 14]
```diff
@@ -753,11 +747,32 @@ static void sanitize_message_from_agent(struct msg_header *untrusted_header)
...
+            if (protocol_version < 3) {
+                fprintf(stderr, "agent sent (new) MSG_TRIGGER_SERVICE3 "
+                        "although it uses protocol %d\n", protocol_version);
+                exit(1);
+            }
+            if (untrusted_header->len < sizeof(struct trigger_service_params3)) {
+                fprintf(stderr, "agent sent invalid MSG_TRIGGER_SERVICE3 packet\n");
+                exit(1);
+            }
+            if (untrusted_header->len-sizeof(struct trigger_service_params3)
```
Comment (HW42) [inline daemon/qrexec-daemon.c]: I would put spaces around the minus.

File: daemon/qrexec-daemon.c [hunk 15]
```diff
@@ -787,7 +806,52 @@ static void handle_message_from_agent(void)
...
+                    &params.request_id);
+            return;
+        case MSG_TRIGGER_SERVICE3:
+            service_name_len = hdr.len - sizeof(untrusted_params3);
+            untrusted_service_name = malloc(service_name_len);
+            if (!untrusted_service_name)
+                handle_vchan_error("malloc(service_name)");
+
+            if (libvchan_recv(vchan, &untrusted_params3, sizeof(untrusted_params3)) < 0)
+                handle_vchan_error("recv params");
```
Comment (HW42) [inline daemon/qrexec-daemon.c]: maybe "recv params3" (same below)

File: libqrexec/qrexec.h [hunk 16]
```diff
@@ -23,9 +23,15 @@
...
+#define QREXEC_PROTOCOL_VERSION 3
 #define MAX_FDS 256
-#define MAX_DATA_CHUNK 4096
+/* protocol version 2 */
+#define MAX_DATA_CHUNK_V2 4096
+/* protocol version 3+ */
+#define MAX_DATA_CHUNK_V3 65536
+
+/* large, but arbitrary */
+#define MAX_SERVICE_NAME_LEN 65536
```
Comment (HW42) [inline libqrexec/qrexec.h]: Probably doesn't matter but why not a bit smaller such that the message will fit one chunk?

File: libqrexec/write-stdin.c [hunk 17]
```diff
@@ -41,8 +41,8 @@ int flush_client_data(int fd, struct buffer *buffer)
         if (!len) {
             return WRITE_STDIN_OK;
         }
-        if (len > MAX_DATA_CHUNK)
-            len = MAX_DATA_CHUNK;
+        if (len > MAX_DATA_CHUNK_V3)
+            len = MAX_DATA_CHUNK_V3;
```
Comment (HW42) [inline libqrexec/write-stdin.c]: What about V2? I guess it's because it's used only on the agent (and you want to support only agent_version < daemon_version) but this looks like a rather generic helper function. So at least add a comment.

File: agent/qrexec-agent-data.c [hunk 18]
```diff
@@ -39,6 +39,10 @@
 
 #define VCHAN_BUFFER_SIZE 65536
 
+#if MAX_DATA_CHUNK_V3 < MAX_DATA_CHUNK_V2
+#error "MAX_DATA_CHUNK_V3 must not be smaller than MAX_DATA_CHUNK_V2"
+#endif
```
[the remark this answers is not in the document]
Comment (marmarek) [reply agent/qrexec-agent-data.c]: I rely on this in `handle_input()` - buffer size is based on `MAX_DATA_CHUNK_V3`, so its assume  to be enough for `MAX_DATA_CHUNK_V2` too. You found it already. But maybe indeed better don't do that at all.

File: agent/qrexec-agent-data.c [hunk 19]
```diff
@@ -152,17 +156,18 @@ void send_exit_code(libvchan_t *data_vchan, int status)
  *  1 - some data processed, call it again when buffer space and more data
  *      available
  */
-int handle_input(libvchan_t *vchan, int fd, int msg_type)
+int handle_input(libvchan_t *vchan, int fd, int msg_type, int data_protocol_version)
 {
-    char buf[MAX_DATA_CHUNK];
-    int len;
+    const size_t max_len = data_protocol_version < 3 ? MAX_DATA_CHUNK_V2 : MAX_DATA_CHUNK_V3;
```
[the remark this answers is not in the document]
Comment (marmarek) [reply agent/qrexec-agent-data.c]: Good idea.

File: daemon/qrexec-client.c (hunk 8 above)
Comment (marmarek) [reply daemon/qrexec-client.c]: I wanted to have it explicitly in `.c` file to guard against changing `.h` without considering all the files using it. Otherwise I'd add just a comment.

File: daemon/qrexec-client.c (hunk 11 above)
Comment (marmarek) [reply daemon/qrexec-client.c]: This isn't strange case. It may mean remote side haven't retrieved previous data yet.

File: libqrexec/qrexec.h (hunk 16 above)
Comment (marmarek) [reply libqrexec/qrexec.h]: Good idea.

File: daemon/qrexec-client.c (hunk 8 above)
Comment (HW42) [reply daemon/qrexec-client.c]: I personally would have added the assert in the header (maybe with a comment where/why this assumption is used/needed). Not only a comment such that the compiler catches a change even if someone didn't read the comment.

Anyway, your version is also reasonable.

As above maybe add a short comment.

File: agent/qrexec-agent-data.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (HW42) [reply agent/qrexec-agent-data.c]: If you keep it, please add a short comment, like "used below to keep the code a bit easier" or something alike. Just to make it clear that why this constraint is introduced here.

File: daemon/qrexec-client.c (hunk 11 above)
Comment (HW42) [reply daemon/qrexec-client.c]: Oh, got confused about the meaning of `libvchan_buffer_space`.

**Comment by HW42** [comment]
How is the version negotiation supposed to work for VM<->VM connections?

If I'm not mistaken we reach
https://github.com/QubesOS/qubes-core-qrexec/blob/d555fba6b23fd7103371cd54b14ed8ae108f0aad/agent/qrexec-agent-data.c#L109-L114
in `qrexec-client-vm` (via `handle_data_client`). So this would break with a domain with old qrexec, no?

**Comment by HW42** [comment]
This seems to be fine, but just curious: Why do we respond with the highest supported protocol version, if we just decided that we need to emulate a older version:
https://github.com/QubesOS/qubes-core-qrexec/blob/d555fba6b23fd7103371cd54b14ed8ae108f0aad/daemon/qrexec-daemon.c#L221

File: libqrexec/write-stdin.c (hunk 17 above)
Comment (marmarek) [reply libqrexec/write-stdin.c]: It doesn't really matter as this is communication with a local process, not remote side over vchan. I'll add a comment.

Comment (marmarek) [reply libqrexec/write-stdin.c]: Hmm, in fact the limit here doesn't make sense at all. If data chunk can't be received by the process as a whole, `write()` will return smaller number. But if it can, sending it in one call will be more performant. I'll remove the check.

**Comment by marmarek** [comment]
> How is the version negotiation supposed to work for VM<->VM connections?

Indeed this is one place I forgot to update.

**Comment by marmarek** [comment]
> This seems to be fine, but just curious: Why do we respond with the highest supported protocol version, if we just decided that we need to emulate a older version:

Mostly to have easier protocol specification and also care less about order of handshake messages. This reduces number of cases to think of to just "each side send its maximum supported version, the lower is used if supported by both sides".

Note that sending negotiated version as a response here may make an *illusion* that it will work with an agent before https://github.com/QubesOS/qubes-core-agent-linux/commit/1aeec025e00a1933c21c181a096a5580406ba8c8, but in practice it isn't true, as it doesn't apply to data channel. So, sticking to this protocol makes it easier to detect such problem.

File: agent/qrexec-agent-data.c (hunk 1 above)
Comment (marmarek) [reply agent/qrexec-agent-data.c]: I'll change that.
A simple test show that additionally heap buffer allocation is significantly faster.

```
[user@devel18 tmp]$ time ./alloc-test # stack

real	0m0.119s
user	0m0.118s
sys	0m0.001s
[user@devel18 tmp]$ time ./alloc-test 1 # heap

real	0m0.004s
user	0m0.002s
sys	0m0.001s
```

<details>

    #include <stdlib.h>
    #include <string.h>

    int alloc1(size_t s) {
        char buf[s];
        memset(buf, 0, sizeof(buf));
    }

    int alloc2(size_t s) {
        char *buf;
        buf = malloc(s);
        memset(buf, 0, sizeof(buf));
        free(buf);
    }

    int main(int argc, char **argv) {
        int i = 100000;

        if (argc == 1) {
            while (i--)
                alloc1(65536);
        } else {
            while (i--)
                alloc2(65536);
        }
    }

**Comment by marmarek** [comment]
@HW42 I've force-pushed changes addressing your review.

**Comment by marmarek** [comment]
@HW42 are the changes ok?

File: daemon/qrexec-client.c [hunk 20]
```diff
@@ -280,18 +286,29 @@ static void send_exit_code(libvchan_t *vchan, int status)
...
+    max_len = libvchan_buffer_space(vchan);
+    if (max_len < sizeof(hdr))
+        return;
+    max_len -= sizeof(hdr);
+    if (max_len > buf_len)
+        max_len = buf_len;
     if (max_len == 0)
         return;
+
+    buf = malloc(buf_len);
```
Comment (HW42) [inline daemon/qrexec-client.c]: Is there a reason for allocating `buf_len` instead of `max_len`? I would call `buf_len`, `data_chunk_size` or something and only allocate `max_len`.

**Comment by HW42** [comment]
> @HW42 are the changes ok?

Mostly, see diff comment.

File: daemon/qrexec-client.c (hunk 20 above)
Comment (marmarek) [reply daemon/qrexec-client.c]: Makes sense. Force-pushed update.

**Comment by marmarek** [comment]
@HW42 is it ok now?
